### DISCUSSION TITLE
Differentiable audio processors w/ the Faust programming language

### DISCUSSION THREAD

**Discussion by DBraun** [jax-ml/jax#13652] [Show and tell]
I made a transpiler from Faust code to JAX. [Faust](https://faust.grame.fr/) is a powerful language for sound synthesis, so connecting it to JAX while supporting differentiability and learnable parameters is a huge research opportunity.

You can explore Faust to JAX in two Colabs. The [first Colab](https://colab.research.google.com/github/DBraun/DawDreamer/blob/main/examples/Faust_to_JAX/Faust_to_JAX.ipynb) uses gradient descent to optimize the cutoff frequency of a low-pass filter. It also shows how to make a differentiable synthesizer like the ones used in Google's [DDSP](https://magenta.tensorflow.org/ddsp) and ByteDance's [Differentiable Wavetable Synthesis](https://arxiv.org/abs/2111.10003). DDSP provided many differentiable modules, but now we can convert the amazing functions in the [Faust Libraries](https://faustlibraries.grame.fr/) to differentiable JAX code. These libraries cover [physical modeling](https://faustlibraries.grame.fr/libs/physmodels/), [ambisonics](https://faustlibraries.grame.fr/libs/hoa/), [filters](https://faustlibraries.grame.fr/libs/filters/), [reverbs](https://faustlibraries.grame.fr/libs/reverbs/), [envelopes](https://faustlibraries.grame.fr/libs/envelopes/), [compressors](https://faustlibraries.grame.fr/libs/compressors/), [wave digital filters](https://faustlibraries.grame.fr/libs/wdmodels/), and much more. Further research could use these processors for [adaptive filters](https://arxiv.org/abs/2204.11942), [audio mastering/FX](https://github.com/adobe-research/DeepAFx-ST), and much more. These differentiable operators can also be integrated with pre-existing JAX model architectures.

The [second Faust to JAX Colab](https://colab.research.google.com/github/DBraun/DawDreamer/blob/main/examples/Faust_to_QDax/Faust_to_QDax.ipynb) uses [QDax](https://github.com/adaptive-intelligent-robotics/QDax/) for quality-diversity learning. In a toy problem, we have a 3-section parametric equalizer with 10 total parameters. We achieve near-perfect fitness in a short amount of time.

If you're in need of business-related use-case involving this project, please let me know 📫 braun$AT$ccrma.stanford.edu
